### PR TITLE
Opt in the null safety

### PR DESCRIPTION
**Pull request by glynskyi**
Opt in the null safety

### CODE REVIEW

**Comment by comigor** [comment]
Hello @glynskyi, thanks for this PR!

I think we still need to wait for all dependencies to be migrated before merging this PR and changing sdk requirements (specially `gql`, to avoid the force non-nullable `!` calls); but this is a great start!

cc @vasilich6107 you were more involved in nnbd, what do you think?

**Comment by vasilich6107** [comment]
@glynskyi thanks for PR.

@comigor we definitely need to check what's  inside)

**Comment by jlnrrg** [comment]
For everyone like me interested in the progress, here is a list of the dependencies listed in pub.dev and their according nullsafety issues:

- [x] [build](https://github.com/dart-lang/build/pull/2996)
- [ ] [code_builder](https://github.com/dart-lang/tools/issues/1136)
- [x] [collection](https://pub.dev/packages/collection/changelog)
- [x] [dart_style](https://github.com/dart-lang/dart_style/issues/978)
- [x] equatable
- [x] [glob](https://github.com/dart-lang/tools/issues/1990) -> done, but will not be merged after experiment flag is not longer required
- [x] [gql and related tools](https://github.com/zino-app/graphql-flutter/issues/797), done but yet to be published on pub.dev
- [x] [http](https://github.com/dart-lang/http/issues/501)
- [x] [json_annotation](https://github.com/google/json_serializable.dart/issues/763)
- [x] [meta](https://pub.dev/packages/meta/changelog)
- [x] [path](https://github.com/dart-lang/path/pull/104)
- [x] recase
- [x] [source_gen](https://github.com/dart-lang/source_gen/pull/520)
- [x] [yaml](https://github.com/dart-lang/yaml/pull/106)

**Comment by featzima** [comment]
Artemis is a code generator and it means that it is not included in the final build (only the generated classes). So it doesn't need to wait until all dependencies are migrates to the null safety. Actually I successfully uses this PR in my null safety projects.

**Comment by simolus3** [comment]
> So it doesn't need to wait until all dependencies are migrates to the null safety

This is not strictly true. Users of code generators can mostly migrate before their `dev_dependencies`, yes. You could probably use the stable version of `artemis` in your null-safety projects too. The `artemis` project itself imports `build`-packages though, so it needs to wait for them to be migrated.

**Comment by jlnrrg** [comment]
@simolus3 you could, but there might be inconsistencies:
Lets take the generated Class GetUserArguments which would accept an `int`. Normally I could pass `null` as a value and everything works as intended, but now my projects variable I pass into is of type `int?` thus there is a missmatch.

**Comment by GP4cK** [comment]
@featzima  how do you do that:
> I successfully uses this PR in my null safety projects.

I tried this in my pubspec.yaml:

```yaml
dev_dependencies:
  artemis:
    git: https://github.com/glynskyi/artemis.git
```

But I get this error:
```
Because every version of artemis from git depends on equatable ^1.2.5 and [your_crappy_project] depends on equatable 2.0.0-nullsafety.3, artemis from git is forbidden.
```

I tried running `flutter clean` and even deleting the pubspec.lock but it didn't fix it...

**Comment by featzima** [comment]
@GP4cK You're able to fix the required library version during the transition period. Just create the `dependency_overrides` section in your `pubspec.yaml` file. The list of required libraries **depends on your dependencies**.  My `pubspec.yaml` looks like:
```
dependency_overrides:
  analyzer: ^0.41.2
  collection: ^1.15.0-nullsafety.4
  rxdart: ^0.24.0
  graphql: ^4.0.0-beta
  gql: '>=0.12.3 <1.0.0'
```

**Comment by jlnrrg** [comment]
@featzima interesting :thinking:. How do you not get the glob issue?
Can you by change post your whole pubspec.yaml
```

Because gql_code_gen 0.1.5 depends on glob ^1.2.0 and no versions of gql_code_gen match >0.1.5 <0.2.0, gql_code_gen ^0.1.5 requires glob ^1.2.0.

And because every version of artemis from git depends on gql_code_gen ^0.1.5, every version of artemis from git requires glob ^1.2.0.

So, because TestProject depends on artemis from git which depends on glob ^2.0.0-nullsafety.0, version solving failed.
pub get failed (1; So, because TestProject depends on artemis from git which depends on glob ^2.0.0-nullsafety.0, version solving failed.)
exit code 1
```

(also forcing glob to version ^2.0.0-nullsafety.0 leads to issues on code generation)

**Comment by featzima** [comment]
@jlnrrg 
<details>
<summary>pubspec.yaml</summary>

```
environment:
  sdk: '>=2.12.0-0.0 <3.0.0'

dependencies:
  flutter:
    sdk: flutter
  cupertino_icons: ^1.0.2
  vector_math: ^2.1.0-nullsafety.5
  path: ^1.8.0-nullsafety.3
  intl: ^0.17.0
  rxdart: ^0.24.0
  sprintf: ^6.0.0-nullsafety
  shared_preferences: ^0.5.12+4
  dio: ^3.0.10
  built_value: ^8.0.0-nullsafety.0
  built_collection: ^5.0.0-nullsafety.0
  http_multi_server: ^2.2.0
  flutter_bloc: ^7.0.0-nullsafety.1
  expandable: ^4.1.4
  scroll_to_index: ^1.0.6
  geolocator: ^7.0.0-nullsafety.5
  geocoding: ^1.0.5
  synchronized: ^3.0.0-nullsafety.1
  vicodin: ^1.0.5-nullsafety.0
  ktx: ^1.1.5-nullsafety
  shimmer: ^2.0.0-nullsafety.0
  sealed_unions: ^3.0.2+2
  flutter_email_sender: ^4.0.0
  url_launcher: ^6.0.0-nullsafety.6
  package_info: ^0.5.0-nullsafety
  flutter_native_timezone: ^1.0.4
  freezed_annotation: ^0.13.0-nullsafety.0
  timezone: ^0.7.0-nullsafety.0
  flinq: ^1.1.1
  trotter: ^1.2.0
  firebase_core: ^0.8.0-1.0.nullsafety.1
  firebase_analytics: ^7.0.1
  firebase_crashlytics: ^0.5.0-1.0.nullsafety.2
  flutter_vibrate: ^1.0.0
  floor: ^1.0.0-nullsafety.0
  rate_app_dialog: ^0.1.2
  google_maps_flutter: ^1.0.6
  equatable: ^2.0.0-nullsafety.3
  app_settings: ^4.0.4
  countdown_flutter: ^0.1.2
  flutter_svg: ^0.20.0-nullsafety.3
  auto_size_text: ^3.0.0-nullsafety.0
  fimber: ^0.5.0-nullsafety.1
  collection: ^1.15.0-nullsafety.4

dependency_overrides:
  analyzer: ^0.41.2
  collection: ^1.15.0-nullsafety.4
  clock: ^1.1.0-nullsafety.3
  intl: ^0.17.0
  path: ^1.8.0
  url_launcher: ^6.0.0-nullsafety.6
  glob: ^2.0.0
  firebase_core: ^0.8.0-1.0.nullsafety.1
  plugin_platform_interface: ^1.1.0-nullsafety.2
  meta: ^1.3.0-nullsafety.6
  ffi: ^0.1.3
  sqlite3: ^0.1.8
  rxdart: ^0.24.0
  graphql: ^4.0.0-beta
  gql: '>=0.12.3 <1.0.0'

dev_dependencies:
  flutter_test:
    sdk: flutter
  build_runner: ^1.11.1
  built_value_generator: ^8.0.0-nullsafety.0
  floor_generator: ^1.0.0-nullsafety.0
  freezed: ^0.13.0-nullsafety.0
  artemis:
    git:
      url: https://github.com/glynskyi/artemis.git
      ref: 68a83f3
  json_serializable: ^4.0.0
```
</details>

**Comment by vasilich6107** [comment]
@jlnrrg meta supports nullsafety from 1.3.0

Scroll this page to the bottom
https://pub.dev/packages/meta/versions

So we can switch on the checkbox

`collection` is from v1.15.0
https://pub.dev/packages/collection/versions

**Comment by LasseRosenow** [comment]
build 2.0.0 with nullsafety was released ;)

@jlnrrg

File: pubspec.yaml [hunk 1]
```diff
@@ -1,40 +1,50 @@
...
+  build: ^1.6.3
+  gql_code_builder: ^0.1.5-alpha+1612972259656
+  gql_dedupe_link: ^2.0.0-nullsafety.1
+  gql_exec: ^0.3.0-nullsafety.1
+  gql_http_link: ^0.4.0-nullsafety.1
+  gql_link: ^0.4.0-nullsafety.1
+  gql: ^0.13.0-nullsafety.1
+  code_builder: ^3.6.0
+  collection: ^1.15.0
+  dart_style: ^1.3.14
```
Comment (Pacane) [inline pubspec.yaml]:
```suggestion
  dart_style: ^2.0.0
```

**Comment by comigor** [comment]
Hey all, we've just released [7.0.0-beta.1](https://pub.dev/packages/artemis/versions/7.0.0-beta.1) as a breaking change, still in beta, version that outputs Artemis generated code as nnbd (eg. GraphQL `Int!` becomes Dart (late) `int`, while GraphQL `Int` becomes Dart `int?`).

Soon I'll start working on migrating Artemis code itself to nnbd, probably rebasing this PR if it's not too much work (given the amount of conflicts – sorry about that, but I lot had to be changed to output code to nnbd), following the dependencies and maybe we can have a full nnbd prerelease.

But please, if your project is already using nnbd, test this new beta and report bugs on top of it! It'd be very helpful to get real use cases (my company has not migrated yet!).

**Comment by jlnrrg** [comment]
@comigor 
I think it would be a good idea to adapt these version too
- yaml: ^3.0.0 (`analyzer >=1.0.0 depends on it`)
- source_gen: ^1.0.0
- glob: ^2.0.0
- dart_style: ^2.0.0
- build: ^2.0.0
- http: ^0.13.0
- gql_link: ^0.4.0-nullsafety.2
- gql_http_link: ^0.4.0-nullsafety.1
- gql_exec: ^0.3.0-nullsafety.1
At this point I stopped comparing step by step and just reference to the [nullsafety version](https://github.com/zino-app/graphql-flutter/blob/nullsafety/packages/graphql/pubspec.yaml) of graphql.
And I also link my [comment](https://github.com/comigor/artemis/pull/259#issuecomment-781898170) again, where the packages with nullsafety are listed

**Comment by inc16sec** [comment]
Is there any progress with this new update (null-safety), still waiting for the packages required by Artemis to be updated.
@jlnrrg have already pointed out most of the packages that need to be updated to the latest release otherwise a build conflict error stops the whole build.

**Comment by comigor** [comment]
@inc16sec Not yet, I didn't have time to adapt the package itself to nnbd. However, since 7.0.0-beta.1 it's possible to [generate nnbd code](https://github.com/comigor/artemis/pull/259#issuecomment-802303948), and technically one could update dependency_overrides as [featzima's comment above](https://github.com/comigor/artemis/pull/259#issuecomment-782607046) and move on (I'd like people to test it and report back, if possible).

**Comment by jlnrrg** [comment]
@comigor thanks for clarifying.
I personally would love to test for errors, but as I highly depend on artemis, I do not want to override so much dependencys with major version changes, as it creates possible conflicts down the line.
also it is a tedious entry barrier for people wanting to test the new nnbd version.

If possible I'd like you to reconsider updating the packages first, thus gaining more testing people 😅

if that is not possible without completely switching to nnbd, then nvm my bad.

**Comment by comigor** [comment]
@jlnrrg Yeah, I'm not waiting for people to test before upgrade, I just didn't had the time to do it yet 😅

**Comment by inc16sec** [comment]
@comigor so I have just tried to override these dependencies:
```yaml
dependency_overrides:
  build: ^2.0.0
  dart_style: '>=1.3.13 <=1.3.13'
  glob: ^2.0.0
  gql: ^0.13.0-nullsafety.2
  gql_exec: ^0.3.0-nullsafety.1
  gql_http_link: ^0.4.0-nullsafety.1
  gql_dedupe_link: ^2.0.0-nullsafety.1
  gql_link: ^0.4.0-nullsafety.2
  graphql: 5.0.0-nullsafety.1
  http: ^0.13.0
  source_gen: ^1.0.0
  yaml: ^3.0.0
 ```
 
And I called `flutter pub run build_runner watch --delete-conflicting-outputs ` but this error showed up.

```
[INFO] Generating build script...
[INFO] Generating build script completed, took 625ms

[INFO] Creating build script snapshot......
[WARNING] stderr: ../../../../documents/flutter/.pub-cache/hosted/pub.dartlang.org/gql_code_gen-0.1.5/lib/ast_builder.dart:83:11: Error: The getter 'name' isn't defined for the class 'DefinitionNode'.
[WARNING] stderr:  - 'DefinitionNode' is from 'package:gql/src/ast/ast.dart' ('../../../../documents/flutter/.pub-cache/hosted/pub.dartlang.org/gql-0.13.0-nullsafety.2/lib/src/ast/ast.dart').
[WARNING] stderr: Try correcting the name to the name of an existing getter, or defining a getter or field named 'name'.
[WARNING] stderr:   if (def.name != null && def.name.value != null) return def.name.value;
[WARNING] stderr:           ^^^^
[WARNING] stderr: ../../../../documents/flutter/.pub-cache/hosted/pub.dartlang.org/gql_code_gen-0.1.5/lib/ast_builder.dart:83:31: Error: The getter 'name' isn't defined for the class 'DefinitionNode'.
[WARNING] stderr:  - 'DefinitionNode' is from 'package:gql/src/ast/ast.dart' ('../../../../documents/flutter/.pub-cache/hosted/pub.dartlang.org/gql-0.13.0-nullsafety.2/lib/src/ast/ast.dart').
[WARNING] stderr: Try correcting the name to the name of an existing getter, or defining a getter or field named 'name'.
[WARNING] stderr:   if (def.name != null && def.name.value != null) return def.name.value;
[WARNING] stderr:                               ^^^^
[WARNING] stderr: ../../../../documents/flutter/.pub-cache/hosted/pub.dartlang.org/gql_code_gen-0.1.5/lib/ast_builder.dart:83:62: Error: The getter 'name' isn't defined for the class 'DefinitionNode'.
[WARNING] stderr:  - 'DefinitionNode' is from 'package:gql/src/ast/ast.dart' ('../../../../documents/flutter/.pub-cache/hosted/pub.dartlang.org/gql-0.13.0-nullsafety.2/lib/src/ast/ast.dart').
[WARNING] stderr: Try correcting the name to the name of an existing getter, or defining a getter or field named 'name'.
[WARNING] stderr:   if (def.name != null && def.name.value != null) return def.name.value;
[WARNING] stderr:                                                              ^^^^
[INFO] Creating build script snapshot... completed, took 10.4s

[SEVERE] Failed to snapshot build script .dart_tool/build/entrypoint/build.dart.
This is likely caused by a misconfigured builder definition.

pub finished with exit code 78
```

And these are the packages that I'm using:

```yaml
dependencies:
  flutter:
    sdk: flutter

  firebase_auth: ^1.0.1
  # freezed_annotation: ^0.14.1
  flutter_bloc: ^7.0.0
  firebase_core: ^1.0.2
  firebase_storage: ^8.0.1

  graphql_flutter: ^5.0.0-nullsafety.1

  flutter_hooks: ^0.16.0
  hooks_riverpod: ^0.13.1+1

  universal_platform: ^1.0.0-nullsafety

  json_serializable: ^4.1.0
  built_value: ^8.0.4
  built_collection: ^5.0.0
  gql: ^0.13.0-nullsafety.2


  # The following adds the Cupertino Icons font to your application.
  # Use with the CupertinoIcons class for iOS style icons.
  cupertino_icons: ^1.0.2
```

**Comment by comigor** [comment]
@inc16sec Hmmm so some breaking changes may be really breaking now :(

**Comment by tlvenn** [comment]
@inc16sec I had the same issue, i got it running for now keeping `gql` with the latest 0.12.x

**Comment by inc16sec** [comment]
> @inc16sec I had the same issue, i got it running for now keeping `gql` with the latest 0.12.x

I have just tried it with `gql: ^0.12.4` and `gql: ` (without specifying a version number) and still got the same error.

**Comment by inc16sec** [comment]
Looks like there's is a problem with`gql_code_gen`.

**Comment by tlvenn** [comment]
I am using this override monstrosity to make it works:

```
dependency_overrides:
  freezed_annotation: 0.14.1
  graphql: 5.0.0-nullsafety.2

  source_gen: ^1.0.0
  glob: ^2.0.0
  dart_style: ^2.0.0
  build: ^2.0.0
  http: ^0.13.0
  http_parser: ^4.0.0
  yaml: ^3.0.0
  # gql: ^0.13.0-nullsafety.2
  gql: ^0.12.4
  gql_exec: ^0.3.0-nullsafety.1
  gql_link: ^0.4.0-nullsafety.2
  gql_http_link: ^0.4.0-nullsafety.1
  gql_dedupe_link: ^2.0.0-nullsafety.1
  gql_build: ^0.2.0-nullsafety.0
  gql_code_builder: ^0.2.0-nullsafety.0
  built_collection: ^5.0.0
  recase: ^4.0.0-nullsafety.0
  graphql_flutter: 5.0.0-nullsafety.1
```

**Comment by inc16sec** [comment]
> I am using this override monstrosity to make it works:
> 
> ```
> dependency_overrides:
>   freezed_annotation: 0.14.1
>   graphql: 5.0.0-nullsafety.2
> 
>   source_gen: ^1.0.0
>   glob: ^2.0.0
>   dart_style: ^2.0.0
>   build: ^2.0.0
>   http: ^0.13.0
>   http_parser: ^4.0.0
>   yaml: ^3.0.0
>   # gql: ^0.13.0-nullsafety.2
>   gql: ^0.12.4
>   gql_exec: ^0.3.0-nullsafety.1
>   gql_link: ^0.4.0-nullsafety.2
>   gql_http_link: ^0.4.0-nullsafety.1
>   gql_dedupe_link: ^2.0.0-nullsafety.1
>   gql_build: ^0.2.0-nullsafety.0
>   gql_code_builder: ^0.2.0-nullsafety.0
>   built_collection: ^5.0.0
>   recase: ^4.0.0-nullsafety.0
>   graphql_flutter: 5.0.0-nullsafety.1
> ```

I can confirm, after overriding the packages you've. provided the builder seemed to work again just fine.
Wondering how?

**Comment by featzima** [comment]
With the latest commit it works without override section in our projects

**Comment by jlnrrg** [comment]
Thanks for the fast support :smile: 
I think this thread will now convert to point out issues :face_with_head_bandage: 
Here I start:
#### replace '@required' with 'required' 
<details><summary>Why is this an issue?</summary>

bc [required is a new keyword](https://github.com/flutter/flutter/issues/67366)

```
lib/model/service/graphql/graphql_api.graphql.dart:95:28: Error: The parameter 'min' can't have a value of 'null' because of its type 'int', but the implicit default value is 'null'.
Try adding either an explicit non-'null' default value or the 'required' modifier.
  IntRange({@required this.min, @required this.max});
                           ^^^

```
</details>

**Comment by comigor** [comment]
Hello everyone, we've just merged https://github.com/comigor/artemis/pull/283 with NNBD support for Artemis (thank you so much @vasilich6107!)

It'll be released as `7.0.0-beta.2`. Test it and please report anything out of place!

Also thanks for your help @glynskyi, but it was easier to migrate without fixing those conflicts.
I'm closing this PR but feel free to open an issue.

**Pull request closed without merging**